src/gui_common/CustomRichTextLabel.cs

```diff
@@ -0,0 +1,245 @@
+/// </summary>
+public class CustomRichTextLabel : RichTextLabel
+{
+    private string extendedBbcode;
+
+    /// <summary>
+    ///   Custom Bbcode tags exclusive for Thrive. Acts more like an extension to the built-in tags.
+    /// </summary>
+    public enum ThriveBbCodeTag
+    {
+        /// <summary>
+        ///   Turns compound internal name string into a bolded display name and an icon next to it.
+        ///   Mainly used in organelle tooltips.
+        /// </summary>
+        Compound,
+
+        /// <summary>
+        ///   Turns input action string into a key prompt image of its primary input event.
+        /// </summary>
+        Input,
+    }
+
+    /// <summary>
+    ///   This supports custom bbcode tags specific to Thrive (for example: [thrive:compound]glucose[/thrive:compound])
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     NOTE: including "thrive" namespace in the tag is a must, otherwise the custom parser wouldn't parse it.
+    ///   </para>
+    /// </remarks>
+    [Export]
+    public string ExtendedBbcode
+    {
+        get => extendedBbcode;
+        set
+        {
+            if (extendedBbcode == value)
+                return;
+
+            extendedBbcode = value;
+
+            // Need to delay this so we can get the correct input controls from settings.
+            Invoke.Instance.Queue(() => ParseCustomTags());
+        }
+    }
+
+    public override void _Ready()
+    {
+        // Make sure bbcode is enabled
+        BbcodeEnabled = true;
+    }
+
+    /// <summary>
+    ///   Parses a custom tagged substring into a templated bbcode.
+    /// </summary>
+    private void ParseCustomTags()
+    {
+        var result = new StringBuilder();
+        var currentTagBlock = new StringBuilder();
+
+        var tagStack = new Stack<string>();
+
+        var isIteratingTag = false;
+        var isIteratingContent = false;
+
+        // The index of a closing bracket in a last iterated opening tag, used
+        // to retrieve the tagged substring
+        var lastStartingTagEndIndex = 0;
+
+        for (int index = 0; index < ExtendedBbcode.Length; index++)
+        {
+            var character = ExtendedBbcode[index];
+
+            var validNextCharacter = index + 1 < ExtendedBbcode.Length && ExtendedBbcode[index + 1] != '[';
+
+            // Opening bracket found, try to parse it
+            if (character == '[' && validNextCharacter && !isIteratingTag &&
+                index < ExtendedBbcode.Length)
+            {
+                // Clear previous tag
+                currentTagBlock.Clear();
+
+                isIteratingTag = true;
+
+                // Skip once, so the bracket doesn't get added into the final string
+                continue;
+            }
+
+            // Character is not a tag, write it normally into the final string
+            if (!isIteratingTag && !isIteratingContent)
+            {
+                result.Append(character);
```

Comment:
I'll let this be, but could you add a TODO comment about trying to add entire substrings at once, as at least in C++ that's the faster approach, instead of going character by character.